### PR TITLE
Issue 22790 variant cache

### LINKED ISSUE #22789: Create Variants Concept into DotCMS
We need to support the `variants` concept into DotCMS so we are going to:

- [x] Create  a new `variants` table: id varchar(36), name varchar(255), archive. We must to update the posgres.sql and mssql.sql file to create this new table Also we need to create a new Upgrade Task
- [x] Create a `Variant` class with to attributes: id and name
- [x] Create `VariantAPI` with methods: save, update, delete, find by id and  archive
- [x] Create `VariantFactory` with methods: save, update, delete, find by id and archive
- [x] Use cache for Variant

### PR DESCRIPTION
**Pull request by freddyucv**
Using cache into VariantFactory

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/variant/business/VariantCacheImpl.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import com.dotcms.util.DotPreconditions;
+import com.dotcms.variant.model.Variant;
+import com.dotmarketing.business.CacheLocator;
+import com.dotmarketing.business.DotCacheAdministrator;
+import com.dotmarketing.business.DotCacheException;
+
+public class VariantCacheImpl implements VariantCache{
+
+    static final String VARIANT_BY_ID = "VARIANT_BY_ID_";
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/variant/business/VariantCacheImpl.java:11]: consider to handle the 404

Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/variant/business/VariantCacheImpl.java:11]: done https://github.com/dotCMS/core/pull/22912/commits/8b4f3aad6a9123079e07d3e2e9a47bfc14d152f1

File: dotCMS/src/main/java/com/dotcms/variant/VariantFactory.java [hunk 2]
```diff
@@ -9,6 +9,12 @@
  */
 public interface VariantFactory {
 
+    Variant VARIANT_404 = Variant.builder()
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/variant/VariantFactory.java:12]: nice

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java [hunk 3]
```diff
@@ -237,13 +252,31 @@ public void updateDeletedField() throws DotDataException {
     public void delete() throws DotDataException {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java:252]: doc

File: dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java [hunk 4]
```diff
@@ -310,33 +351,46 @@ public void getArchived() throws DotDataException {
     /**
      * Method to test: {@link VariantFactory#get(String)}
      * When: Try to get  {@link Variant} by id that not exists
-     * Should: return a {@link Optional#empty()}
+     * Should:
+     * - return a {@link Optional#empty()}
+     * - Storage as {@link VariantFactory#VARIANT_404} in the cache
+     * - Return null if the {@link VariantFactory#get(String)} is called twice
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java:357]: return null?

File: dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java [hunk 5]
```diff
@@ -310,33 +351,46 @@ public void getArchived() throws DotDataException {
...
 
     /**
      * Method to test: {@link VariantFactory#get(String)}
      * When: Try to get  {@link Variant} by name that not exists
-     * Should: return a {@link Optional#empty()}
-     *
+     * Should:
+     * - return a {@link Optional#empty()}
+     * - Storage as {@link VariantFactory#VARIANT_404} in the cache
+     * - Return null if the {@link VariantFactory#get(String)} is called twice
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java:380]: null?

File: dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java (hunk 4 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java]: https://github.com/dotCMS/core/pull/22912/commits/8625594d00ae40ea121f9c24e8d8c1155c641096

File: dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java (hunk 5 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/variant/VariantFactoryTest.java]: done https://github.com/dotCMS/core/pull/22912/commits/8625594d00ae40ea121f9c24e8d8c1155c641096

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/variant/VariantFactoryImpl.java [hunk 6]
```diff
@@ -80,37 +87,61 @@ public void update(Variant variant) throws DotDataException {
                 .addParam(variant.archived())
                 .addParam(variant.identifier())
                 .loadResult();
+
+        CacheLocator.getVariantCache().put(variant);
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/variant/VariantFactoryImpl.java:91]: We need to invalidate when a variant is updated, not put.  We only send invalidations around in a cluster.  Loading from cache always needs to be done lazily.

File: dotCMS/src/main/java/com/dotcms/variant/VariantFactoryImpl.java [hunk 7]
```diff
@@ -80,37 +87,61 @@ public void update(Variant variant) throws DotDataException {
...
 
     @Override
     public Optional<Variant> get(final String identifier) throws DotDataException {
-        final ArrayList loadResults = new DotConnect().setSQL(VARIANT_SELECT_QUERY)
-                .addParam(identifier)
+        Variant variant = CacheLocator.getVariantCache().getById(identifier);
+
+        return UtilMethods.isSet(variant) && !variant.equals(VARIANT_404) ?
+                Optional.of(variant) :
+                getFromDataBaseById(identifier).or(() -> {
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/variant/VariantFactoryImpl.java:112]: we should be adding the found variant to the cache.

**Pull request merged**